rust/lance/src/index/vector/ivf.rs

```diff
@@ -969,112 +979,82 @@ pub async fn build_ivf_hnsw_index(
-    async fn load_and_remap(
-        mut self,
-        reader: Arc<dyn Reader>,
-        index: &IVFIndex,
-        mapping: &HashMap<u64, Option<u64>>,
-    ) -> Result<Self> {
-        let mut page = index
-            .sub_index
-            .load(reader, self.offset, self.length as usize)
-            .await?;
-        page.remap(mapping)?;
-        self.page = Some(page);
-        Ok(self)
-    }
-
-    async fn write(self, writer: &mut ObjectWriter, ivf: &mut Ivf) -> Result<()> {
-        let page = self.page.as_ref().expect("Load was not called");
-        let page: &PQIndex = page
-            .as_any()
-            .downcast_ref()
-            .expect("Generic index writing not supported yet");
-        ivf.offsets.push(writer.tell().await?);
-        ivf.lengths
-            .push(page.row_ids.as_ref().unwrap().len() as u32);
-        PlainEncoder::write(writer, &[page.code.as_ref().unwrap().as_ref()]).await?;
-        PlainEncoder::write(writer, &[page.row_ids.as_ref().unwrap().as_ref()]).await?;
-        Ok(())
-    }
-}
-
-fn generate_remap_tasks(offsets: &[usize], lengths: &[u32]) -> Result<Vec<RemapPageTask>> {
-    let mut tasks: Vec<RemapPageTask> = Vec::with_capacity(offsets.len() * 2 + 1);
-
-    for (offset, length) in offsets.iter().zip(lengths.iter()) {
-        tasks.push(RemapPageTask::new(*offset, *length));
-    }
-
-    Ok(tasks)
-}
-
 #[allow(clippy::too_many_arguments)]
-pub(crate) async fn remap_index_file(
+pub(crate) async fn mutate_index_file(
     dataset: &Dataset,
-    old_uuid: &str,
     new_uuid: &str,
     old_version: u64,
     index: &IVFIndex,
-    mapping: &HashMap<u64, Option<u64>>,
     name: String,
     column: String,
     transforms: Vec<pb::Transform>,
+    mut mutate_pq: impl FnMut(&mut PQIndex) -> Result<()>,
+    mut mutate_ivf: impl FnMut(&mut Ivf) -> Result<()>,
 ) -> Result<()> {
+    // TODO: make this function work for other types of indices.
     let object_store = dataset.object_store();
-    let old_path = dataset.indices_dir().child(old_uuid).child(INDEX_FILE_NAME);
+    let old_path = dataset
+        .indices_dir()
+        .child(index.uuid.as_str())
+        .child(INDEX_FILE_NAME);
     let new_path = dataset.indices_dir().child(new_uuid).child(INDEX_FILE_NAME);
 
     let reader: Arc<dyn Reader> = object_store.open(&old_path).await?.into();
     let mut writer = object_store.create(&new_path).await?;
 
-    let tasks = generate_remap_tasks(&index.ivf.offsets, &index.ivf.lengths)?;
-
-    let mut task_stream = stream::iter(tasks.into_iter())
-        .map(|task| task.load_and_remap(reader.clone(), index, mapping))
-        .buffered(num_cpus::get());
+    let offsets = index.ivf.offsets.as_slice();
+    let lengths = index.ivf.lengths.as_slice();
+    let mut page_stream = futures::stream::iter(offsets.iter().zip(lengths.iter()))
+        .map(|(offset, length)| {
+            index
+                .sub_index
+                .load(reader.clone(), *offset, *length as usize)
```

Comment:
I'm working on new index format, the semantics of `load` now is loading the index, and `load_partition` is loading the index as a sub index.
the latter requires `partition_id` to load the sub index with partition metadata of given partition.